src/sagemaker/tensorflow/legacy_training.rst

```diff
@@ -0,0 +1,487 @@
+        "rmse": tf.metrics.root_mean_squared_error(tf.cast(labels, tf.float64), predictions)
+    }
+
+    optimizer = tf.train.GradientDescentOptimizer(
+        learning_rate=hyperparameters["learning_rate"])
+    train_op = optimizer.minimize(
+        loss=loss, global_step=tf.train.get_global_step())
+
+    # Provide an estimator spec for `ModeKeys.EVAL` and `ModeKeys.TRAIN` modes.
+    return tf.estimator.EstimatorSpec(
+        mode=mode,
+        loss=loss,
+        train_op=train_op,
+        eval_metric_ops=eval_metric_ops)
+
+Distributed training
+''''''''''''''''''''
+
+When distributed training happens, a copy of the same neural network will be sent to
+multiple training instances. Each instance will train with a batch of the dataset,
+calculate loss and minimize the optimizer. One entire loop of this process is called training step.
+
+A `global step <https://www.tensorflow.org/api_docs/python/tf/train/global_step>`_ is a global
+counter shared between the instances. It is necessary for distributed training, so the optimizer
+can keep track of the number of training steps across instances. The only change in the
+previous complete ``model_fn`` to enable distributed training is to pass in the global
+step into the ``optimizer.minimize`` function:
+
+.. code:: python
+
+  train_op = optimizer.minimize(loss, tf.train.get_or_create_global_step())
+
+More information about distributed training can be find in talk from the TensorFlow Dev Summit 2017
+`Distributed TensorFlow <https://www.youtube.com/watch?time_continue=1&v=la_M6bCV91M>`_.
+
+
+More details on how to create a ``model_fn`` can be find in `Constructing the model_fn <https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/docs_src/extend/estimators.md#constructing-the-model_fn-constructing-modelfn>`_.
+
+
+Creating ``train_input_fn`` and ``eval_input_fn`` functions
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The ``train_input_fn`` is used to pass ``features`` and ``labels`` to the ``model_fn``
+in **training** mode. The ``eval_input_fn`` is used to ``features`` and ``labels`` to the
+``model_fn`` in **evaluation** mode.
+
+The basic skeleton for the ``train_input_fn`` looks like this:
+
+.. code:: python
+
+  def train_input_fn(training_dir, hyperparameters):
+    # Logic to the following:
+    # 1. Reads the **training** dataset files located in training_dir
+    # 2. Preprocess the dataset
+    # 3. Return 1)  a dict of feature names to Tensors with
+    # the corresponding feature data, and 2) a Tensor containing labels
+    return features, labels
+
+An ``eval_input_fn`` follows the same format:
+
+.. code:: python
+
+  def eval_input_fn(training_dir, hyperparameters):
+    # Logic to the following:
+    # 1. Reads the **evaluation** dataset files located in training_dir
+    # 2. Preprocess the dataset
+    # 3. Return 1)  a dict of feature names to Tensors with
+    # the corresponding feature data, and 2) a Tensor containing labels
```

Comment:
same thing in this section - I find it a bit weird to have the steps in comments